tools/themis/src/utils.rs

```diff
@@ -0,0 +1,123 @@
+            let raw = parse_toml(&package.manifest_path)?;
+            Ok(Package { parsed: package, raw })
+        })
+        .collect::<anyhow::Result<Vec<Package>>>()?;
+
+    Ok(Workspace { root: metadata.workspace_root, members })
+}
+
+/// returns true if the rule failed, false if otherwise
+pub fn check_and_report(outcome: Result<(), Error>, workspace: &Workspace) -> anyhow::Result<bool> {
+    match outcome {
+        Err(Error::RuntimeError(err)) => Err(err),
+        Err(Error::OutcomeError { msg, expected, outliers }) => {
+            let mut report = format!(
+                "{c_heading}(i) {}:{c_none}{}",
+                msg,
+                expected.map_or("".to_string(), |Expected { value, reason }| format!(
+                    " [expected: {c_expected}{}{c_none}{}]",
+                    value,
+                    reason.map_or("".to_string(), |reason| format!(", {}", reason)),
+                    c_expected = style::fg(style::Color::Color256(35))
+                        + &style::bg(style::Color::Gray { shade: 3 })
+                        + style::bold(),
+                    c_none = style::reset()
+                )),
+                c_heading = style::fg(style::Color::Color256(172)) + style::bold(),
+                c_none = style::reset()
+            );
+
+            for PackageOutcome { pkg: Package { parsed: pkg, .. }, value } in outliers {
+                report.push_str(&format!(
+                    "\n \u{2022} {c_name}{}{c_none} v{} {c_path}({}){c_none}{}",
+                    pkg.name,
+                    pkg.version,
+                    pkg.manifest_path.strip_prefix(&workspace.root).unwrap(),
+                    value.map_or("".to_string(), |v| format!(
+                        " [found: {c_found}{}{c_none}]",
+                        v,
+                        c_found = style::fg(style::Color::White)
+                            + &style::bg(style::Color::Red)
+                            + style::bold(),
+                        c_none = style::reset()
+                    )),
+                    c_name = style::fg(style::Color::Color256(39)) + style::bold(),
+                    c_path = style::fg(style::Color::Gray { shade: 12 }),
+                    c_none = style::reset(),
+                ));
+            }
+            eprintln!("{}", report);
+            Ok(true)
+        }
+        _ => Ok(false),
+    }
+}
+
+pub fn is_publishable(pkg: &Package) -> bool {
+    !matches!(pkg.raw["package"].get("publish"), Some(toml::Value::Boolean(false)))
+}
+
+/// Returns true if a crate has either of the provided files.
+///
+/// ## Example
+///
+/// ```rust
+/// exists!(pkg, "a" || "b" || ( "c" && "d" ) || "e")
+/// ```
+///
+/// If the crate has either `a`, `b`, `e` or both `c` and `d`, this returns true.
+/// Only returns false if all the combinations don't exist.
+macro_rules! exists {
+    ($pkg:expr, $($files:tt)+) => {{
+        let pkg_root = $pkg.parsed.manifest_path.parent().unwrap();
```

Comment:
Maybe, it's worth changing macro, in such a way that it avoid hidden `unwrap().`

And instead returns `Result<bool, Error>`.

This is not  blocking, though it would be less surprising to the reader, that the function can panic.

Reply:
Well, this actually isn't ever going to panic. Here's why:

For any crate that exists, the manifest path would be `<pkg>/Cargo.toml`. the parent, would be `<pkg>`

We know, for a fact that there would always be a parent if we even have a crate to begin with.

And, though unlikely, let's game this out and say, what if the crate exists on the root directory. In which case, for `/Cargo.toml`, the parent directory would then be `/`

This unwrap call will never panic and the guarantee is that there's a crate here at all to begin with.

https://docs.rs/camino/latest/camino/struct.Utf8Path.html#method.parent

Reply:
@miraclx I understand that it can't panic, and why. My point was that when reading the code, whenever I see `unwrap()` I need to spend a moment and thing about. Instead, adding a comment "can't panic", or writing code in such a way that avoids "unwrap()" would be better.

I'm not saying you have to change anything, just something to thinking about. This is a script, not a production code, so it's fine.